scripts/registries-test.py

```diff
@@ -0,0 +1,268 @@
+#!/usr/bin/env python3
+"""
+Test `uv add` against multiple Python package registries.
+
+This script looks for environment variables that configure registries for testing.
+To configure a registry, set the following environment variables:
+
+    UV_TEST_<registry_name>_URL         URL for the registry
+    UV_TEST_<registry_name>_TOKEN       authentication token
+    UV_TEST_<package_name>_PKG          private package to install
+
+The username defaults to "__token__" but can be optionally set with:
+    UV_TEST_<registry_name>_USERNAME
+
+Keep in mind that some registries can fall back to PyPI internally, so make sure
+you choose a package that only exists in the registry you are testing.
+
+# /// script
+# dependencies = ["colorama"]
+# ///
+"""
+
+import argparse
+import os
+import re
+import subprocess
+import sys
+import tempfile
+from pathlib import Path
+from typing import Dict
+
+from colorama import Fore
+from colorama import init as colorama_init
+
+colorama_init(autoreset=True)
+
+DEFAULT_TIMEOUT = 30
+
+KNOWN_REGISTRIES = [
+    "artifactory",
+    "azure",
+    "aws",
+    "cloudsmith",
+    "gcp",
+    "gemfury",
+    "gitlab",
+    "nexus",
+]
+
+
+def get_registries() -> Dict[str, str]:
+    pattern = re.compile(r"^UV_TEST_(.+)_URL$")
+    registries: Dict[str, str] = {}
+
+    for env_var, value in os.environ.items():
+        match = pattern.match(env_var)
+        if match:
+            registry_name = match.group(1).lower()
+            registries[registry_name] = value
+
+    return registries
+
+
+def setup_test_project(registry_name: str, registry_url: str, project_dir: str) -> Path:
+    """Create a temporary project directory with a pyproject.toml"""
+    pyproject_content = f"""[project]
+name = "{registry_name}-test"
+version = "0.1.0"
+description = "Test registry"
+
+[[tool.uv.index]]
+name = "{registry_name}"
+url = "{registry_url}"
+default = true
+"""
+
+    pyproject_file = Path(project_dir) / "pyproject.toml"
+    pyproject_file.write_text(pyproject_content, encoding="utf-8")
+
+
+def run_test(
+    registry_name: str,
+    registry_url: str,
+    package: str,
+    username: str,
+    token: str,
+    verbosity: int,
+    timeout: int = DEFAULT_TIMEOUT,
+) -> bool:
+    """Attempt to install package from this registry."""
+    print(
+        f"{registry_name} -- Running test for {registry_url} with username {username}"
+    )
+    print(f"\nAttempting to install {package}")
+    os.environ[f"UV_INDEX_{registry_name.upper()}_USERNAME"] = username
+    os.environ[f"UV_INDEX_{registry_name.upper()}_PASSWORD"] = token
+
+    with tempfile.TemporaryDirectory() as project_dir:
+        cmd = [
+            "cargo",
+            "run",
+            "--",
```

Comment:
It'd probably be good to just use `"uv"` and accept a path to the uv binary to use. It'd be nice not to entangle this with `cargo build` / `cargo run` so we can test against other versions and such,